cli: colocate stories and components, centralize main.js

tooppaaa opened the pull request:
Issue: #10901

## What I did

- Move components from @storybook/< framework >/demo to cli/frameworks/react/js and /ts
- Copy stories folder to user src (if any) or to root
- mutualise generator

## How to test
 
- E2E should be green
- use CLI

tooppaaa committed:
rename generator

tooppaaa committed:
CLI: react stories

tooppaaa committed:
fix build

tooppaaa committed:
fixes

tooppaaa committed:
fix tests

tooppaaa committed:
fix unit test

tooppaaa committed:
fix lint

tooppaaa committed:
fix tests

tooppaaa committed:
fix unit test

shilman commented on lib/cli/src/frameworks/react/js/0-Welcome.stories.js:
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import { linkTo } from '@storybook/addon-links';
-import { Welcome } from '@storybook/react/demo';
+import Welcome from './Welcome';
```
Any rationale for default vs named imports? I ask bc we're standardizing on named imports in Storybook and apparently a lot of people feel passionately about that.

tooppaaa replied:
I'm an export default person :D no worries of changing it

shilman commented on lib/cli/src/frameworks/react/ts/Welcome.tsx:
```diff
@@ -0,0 +1,159 @@
+  HTMLAttributes,
+} from 'react';
+
+type MainProps = Omit<DetailedHTMLProps<HTMLAttributes<HTMLElement>, HTMLElement>, 'style'>;
```
Can we simplify the TypeScript in our welcome component? Otherwise it's very welcoming!

tooppaaa replied:
I have in mind of having Welcome being a mdx story and having another more simple/friendly component.  Not sure how to simplify this one

gaetanmaisse replied:
Agree with @shilman a simpler type would be great but I'm a React newbie so I don't if we can do better or if there is a workaround.

shilman commented on lib/cli/src/frameworks/react/ts/Welcome.tsx:
```diff
@@ -0,0 +1,159 @@
+interface WelcomeProps {
+  showApp: () => void;
+}
+const Welcome: FunctionComponent<WelcomeProps> = ({ showApp }) => (
```
Can we add `{ showApp }: WelcomeProps` here? It's currently required if the user switches to `react-docgen` for props tables. It's not the default, but I expect lots of users will try switching and immediately the props will disappear without that type annotation due to https://github.com/reactjs/react-docgen/issues/387

shilman commented on lib/cli/src/helpers.test.ts:
```diff
@@ -54,4 +56,58 @@ describe('Helpers', () => {
   });
+
+  it.each`
+    language        | exists          | expected
```
WHOA!!! 😍

shilman reviewed (commented):
This looks great @tooppaaa! Some small nitpicks in the review, but otherwise feels good.

tooppaaa committed:
Feedbacks

tooppaaa commented on lib/cli/src/generators/configure.ts:
```diff
@@ -0,0 +1,19 @@
+  const hasSrc = fse.existsSync('./src');
+
+  const config = {
+    stories: [!hasSrc && '../stories/**/*.stories.*', hasSrc && '../src/**/*.stories.*'].filter(
```
@ndelangen How does this fit with your fix on globs ?

ndelangen replied:
those are valid globs AFAICS

tooppaaa committed:
Move ember

tooppaaa committed:
Move Meteor

tooppaaa committed:
Move mithril

tooppaaa committed:
Move preact

tooppaaa committed:
Move rax

tooppaaa committed:
Move riot

tooppaaa committed:
Move svelte

tooppaaa committed:
Move marko and cleanup

tooppaaa committed:
Move marionette

tooppaaa committed:
Move vue

tooppaaa committed:
Move html

ndelangen committed:
no tabs

tooppaaa committed:
Merge remote-tracking branch 'origin/next' into tech/colocateReactStories

tooppaaa committed:
Merge branch 'tech/colocateReactStories' into tech/colocateSomeStories

tooppaaa committed:
Merge pull request #10935 from storybookjs/tech/colocateSomeStories

Tech/colocate some stories

tooppaaa committed:
Move WebComponents

tooppaaa committed:
Move react-native

tooppaaa committed:
Move & clean aurelia

tooppaaa committed:
Move and clean Angular

tooppaaa committed:
Oops

tooppaaa committed:
fix tests

tooppaaa committed:
Merge remote-tracking branch 'origin/next' into tech/colocateReactStories

tooppaaa committed:
Merge branch 'tech/colocateReactStories' into tech/colocateMoreStories

tooppaaa committed:
Merge pull request #10972 from storybookjs/tech/colocateMoreStories

Tech/colocate more stories

tooppaaa committed:
add angular stories back

tooppaaa committed:
fix aurelia

tooppaaa committed:
fix welcome story

tooppaaa committed:
remove plugin-external-helpers from fixtures

tooppaaa commented:
As per the old https://github.com/storybookjs/storybook/issues/1320 I removed the plugin from the fixture. I'm not sure why the fixture failed to display the Welcome component.

shilman commented on CHANGELOG.md:
```diff
@@ -1,3 +1,35 @@
+## 6.0.0-beta.17 (May 28, 2020)
```
???

tooppaaa replied:
hmm, probably bad merge

tooppaaa committed:
Merge remote-tracking branch 'origin/next' into tech/colocateReactStories

tooppaaa committed:
merge

tooppaaa committed:
fix copy template when needed

tooppaaa committed:
fix build

tooppaaa committed:
switch to docs over essentials

tooppaaa committed:
fix build

tooppaaa committed:
make it green

tooppaaa commented:
https://github.com/storybookjs/storybook/issues/9103 is also impacting this as it's the reason of this nice line of code: `framework !== 'angular' && '@storybook/addon-essentials',`

I'm hoping removing it before being able to merge this.
@shilman any clue on the chromatic diff ? 

Before cleaning up here's what's left:
- [ ] Using SFC for Vue
- [ ] Test react
- [ ] Test react-native
- [ ] Test vue
- [ ] Test angular 
- [ ] Test mithril
- [ ] Test riot
- [ ] Test ember
- [ ] Test marionette
- [ ] Test marko
- [ ] Test meteor
- [ ] Test preact
- [ ] Test svelte
- [ ] Test rax
- [ ] Test aurelia
- [ ] Test html
- [ ] Test web-components

I'm confident to quickly have a PR with a single Welcome.mdx to cleanup template-mdx so I won't do much tests, but they should work anyway

tooppaaa committed:
Fix yarn 2

gaetanmaisse commented on lib/cli/src/helpers.ts:
```diff
@@ -344,3 +345,42 @@ export function copyTemplate(templateRoot: string, storyFormat: StoryFormat) {
 }
+
+export function copyComponents(framework: SupportedFrameworks, language: SupportedLanguage) {
+  const languageFolderMapping: { [key in SupportedLanguage]: string } = {
```
`{ [key in SupportedLanguage]: string }` can be replaced by TS `Record` builtin type: `Record<SupportedLanguage, string>`

tooppaaa replied:
I always forgot about it. Thanks !

gaetanmaisse commented on lib/cli/src/generators/generator.ts:
```diff
@@ -0,0 +1,80 @@
+  addScripts: true,
+  addComponents: true,
+};
+const generator = async (
```
Maybe it can be named `baseGenerator` and exported directly (I mean `export const baseGenerator =`)?

gaetanmaisse reviewed (approved):
Now let's take a look at this pb with Yarn 2 😄 🙃

gaetanmaisse commented on lib/cli/src/generators/configure.ts:
```diff
@@ -0,0 +1,24 @@
+    addons: [
+      '@storybook/addon-actions',
+      '@storybook/addon-links',
+      '@storybook/addon-docs',
```
Shouldn't it be `@storybook/addon-essentials` instead?

gaetanmaisse replied:
See https://github.com/storybookjs/storybook/pull/10913/commits/2570a66ae08d16e6110823627878258740785a96

gaetanmaisse commented:
Yarn 2 issue seems related to this line: https://github.com/storybookjs/storybook/blob/next/lib/core/src/server/presets.js#L124

gaetanmaisse committed:
fix(cli): use the same list of addons to build `main.js` as the one added during init

gaetanmaisse committed:
fix(cli): add peer deps of storybook addons when init a project

Some addons have peer deps that must be added to make everything works, like '@storybook/addon-docs' => 'react-is'

gaetanmaisse commented on lib/cli/src/generators/generator.ts:
```diff
@@ -50,9 +50,20 @@ const generator = async (
-    Boolean
-  );
+  // ⚠️ Some addons have peer deps that must be added too, like '@storybook/addon-docs' => 'react-is'
+  const addonsPeerDeps = addons.some(
+    (addon) => addon === '@storybook/addon-essentials' || addon === '@storybook/addon-docs'
+  )
+    ? ['react-is']
+    : [];
```
@shilman any reason to have `react-is` as a peerDep and not a regular dep? I'm guessing because having multiples different versions in a single project leads to chaos💥?

tooppaaa committed:
PR Feedbacks

tooppaaa committed:
Use docs & actions for now

tooppaaa commented on lib/cli/test/fixtures/webpack_react_frozen_lock/.babelrc:
```diff
@@ -1,4 +1,3 @@
 {
-  "presets": ["@babel/preset-react"],
```
Revert

tooppaaa commented on lib/cli/test/fixtures/webpack_react_frozen_lock/package.json:
```diff
@@ -12,7 +12,6 @@
   },
   "devDependencies": {
     "@babel/core": "^7.3.4",
-    "@babel/plugin-external-helpers": "7.2.0",
```
revert

ndelangen committed:
Merge branch 'next' into tech/colocateReactStories

tooppaaa committed:
Revert "remove plugin-external-helpers from fixtures"

This reverts commit fde26b373f2ef0f02d4012b144d57ca9029b5db5.

tooppaaa committed:
move vue

tooppaaa committed:
Merge remote-tracking branch 'origin/next' into tech/colocateReactStories

tooppaaa committed:
finish merge

tooppaaa committed:
fix merge

tooppaaa commented:
Closing in favor of #11136
